src/formatting/newline_style.rs

```diff
@@ -0,0 +1,181 @@
+use crate::NewlineStyle;
+
+/// Apply this newline style to the formatted text. When the style is set
+/// to `Auto`, the `raw_input_text` is used to detect the existing line
+/// endings.
+///
+/// If the style is set to `Auto` and `raw_input_text` contains no
+/// newlines, the `Native` style will be used.
+pub(crate) fn apply_newline_style(
+    newline_style: NewlineStyle,
+    formatted_text: &mut String,
+    raw_input_text: &str,
+) {
+    match effective_newline_style(newline_style, raw_input_text) {
+        EffectiveNewlineStyle::Windows => {
+            *formatted_text = convert_to_windows_newlines(formatted_text);
+        }
+        EffectiveNewlineStyle::Unix => {}
+    }
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+enum EffectiveNewlineStyle {
+    Windows,
+    Unix,
+}
+
+fn effective_newline_style(
+    newline_style: NewlineStyle,
+    raw_input_text: &str,
+) -> EffectiveNewlineStyle {
+    match newline_style {
+        NewlineStyle::Auto => auto_detect_newline_style(raw_input_text),
+        NewlineStyle::Native => native_newline_style(),
+        NewlineStyle::Windows => EffectiveNewlineStyle::Windows,
+        NewlineStyle::Unix => EffectiveNewlineStyle::Unix,
+    }
+}
+
+const LINE_FEED: char = '\n';
+const CARRIAGE_RETURN: char = '\r';
+
+fn auto_detect_newline_style(raw_input_text: &str) -> EffectiveNewlineStyle {
+    let first_line_feed_pos = raw_input_text.chars().position(|ch| ch == LINE_FEED);
+    match first_line_feed_pos {
+        Some(first_line_feed_pos) => {
+            let char_before_line_feed_pos = first_line_feed_pos.saturating_sub(1);
+            let char_before_line_feed = raw_input_text.chars().nth(char_before_line_feed_pos);
+            match char_before_line_feed {
+                Some(CARRIAGE_RETURN) => EffectiveNewlineStyle::Windows,
+                _ => EffectiveNewlineStyle::Unix,
+            }
+        }
+        None => native_newline_style(),
+    }
+}
+
+fn native_newline_style() -> EffectiveNewlineStyle {
+    if cfg!(windows) {
+        EffectiveNewlineStyle::Windows
+    } else {
+        EffectiveNewlineStyle::Unix
+    }
+}
+
+fn convert_to_windows_newlines(formatted_text: &String) -> String {
+    let mut transformed = String::with_capacity(2 * formatted_text.capacity());
+    for c in formatted_text.chars() {
+        const WINDOWS_NEWLINE: &str = "\r\n";
+        match c {
+            LINE_FEED => transformed.push_str(WINDOWS_NEWLINE),
```

Comment:
may it happen that for an input text like `aaa\r\nbbb` you get `aaa\r\r\nbbb`, i.e., the extra `\r` character ?

Reply:
You're absolutely right I didn't think of that! I added tests for that and fixed it in 1f09d94.